### PR TITLE
[6.0] Make simple color field a proper web component

### PR DESCRIPTION
**Pull request by dgrammatiko**
### Summary of Changes

Refactor the code so:
- php: layout spits buttons as slotted elements
- use of [:part](https://developer.mozilla.org/en-US/docs/Web/CSS/::part) to style the component internals.
- component becomes light/dark
- JS: component now uses the [elementInternals](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals) so it is an **actual** form field
- The client side supports all the colors that a browser could support (hex, rgb, lch, oklab, `color(colorspace val1 val2 val3 / alpha)`). This requires setting the format to `""`. All color values are validated client side

### B/C considerations:
- The field layout has changes (the old one WILL NOT WORK)
- The old field dispatched 2 change events, one on the custom element and another on the select element. There's only one event on the custom element now

### Testing Instructions

Paste this:

```xml
<field
  name="blabla"
  type="color"
  control="simple"
  format=""
  colors="#000000,rgb(59% 4.79% 75.2% / 0.8),oklch(87% 0.293 207),oklab(79% -0.261 0.088),color(display-p3 0.83 0.19 0.73 / 0.96),#FF0000,lch(56 90.2 154 / 0.88),none"
  default=""
  label="color"
/>
```

in the `templateDetails.xml` of the Atum template (after line 51: `<fieldset name="colors" label="TPL_ATUM_COLORS_SETTINGS_LABEL">`)

Check that the field:
- saves and respects the value from the PHP side
- works with the keyboard (enter to go to the select mode, tab to navigate, enter to select)

### Expected result AFTER applying this Pull Request

https://github.com/joomla/joomla-cms/assets/3889375/d2753f25-be63-4548-a0c4-b765c4579e83

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [x] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

#### Unused strings

```
JFIELD_COLOR_SELECT="Select a colour"
JFIELD_COLOR_TRANSPARENT="No colour, transparent"
JFIELD_COLOR_VALUE="Colour with hexadecimal value of"
```

### CODE REVIEW

File: administrator/language/en-GB/joomla.ini [hunk 1]
```diff
@@ -219,9 +219,6 @@ JFIELD_COLOR_LABEL_SLIDER_HUE="Hue Slider"
 JFIELD_COLOR_LABEL_SLIDER_INPUT="Selected Colour Value"
 JFIELD_COLOR_LABEL_SLIDER_LIGHT="Light Slider"
 JFIELD_COLOR_LABEL_SLIDER_SATURATION="Saturation Slider"
-JFIELD_COLOR_SELECT="Select a colour"
```
Comment (brianteeman) [inline administrator/language/en-GB/joomla.ini:222]: the policy is that they must be deprecated before they can be removed

Comment (dgrammatiko) [reply administrator/language/en-GB/joomla.ini:222]: Ok, is 5.x deprecation good enough or there's some time/version limit (ie 2 versions)?

File: layouts/joomla/form/field/color/simple.php [hunk 2]
```diff
@@ -51,22 +52,34 @@
...
     ->useScript('webcomponent.field-simple-color');
 
+Text::script('JCLOSE');
+Text::script('JNONE');
+
+$slots = [];
+$attr  = [
+    'name'  => $name,
+    'id'    => $id,
+    'class' => '' . trim($class),
```
Comment (Quy) [inline layouts/joomla/form/field/color/simple.php]:
```suggestion
    'class' => trim($class),
```
?

Comment (dgrammatiko) [reply layouts/joomla/form/field/color/simple.php]: I guess copy paste...

Thanks

File: layouts/joomla/form/field/color/simple.php [hunk 3]
```diff
@@ -51,22 +52,34 @@
...
+
+if ($disabled) {
+    $attr['disabled'] = '';
+}
+
+if ($readonly) {
+    $attr['readonly'] = '';
+}
+
+foreach ($colors as $key => $val) {
```
Comment (Quy) [inline layouts/joomla/form/field/color/simple.php]:
```suggestion
foreach ($colors as $val) {
```

**Comment by Quy** [comment]
Opening the color picker shifts up the `Choose your hue...` label slightly.

**Comment by dgrammatiko** [comment]
> Opening the color picker shifts up the Choose your hue... label slightly.

What's your browser?

**Comment by Quy** [comment]
Windows Firefox 124.0. It is not an issue with Chrome.

File: administrator/language/en-GB/joomla.ini (hunk 1 above)
Comment (Quy) [reply administrator/language/en-GB/joomla.ini:222]: Please deprecate to be removed in v7.0.

File: layouts/joomla/form/field/color/simple.php (hunk 3 above)
Comment (Quy) [reply layouts/joomla/form/field/color/simple.php]: This got reverted in the last commit.

File: administrator/language/en-GB/joomla.ini (hunk 1 above)
Comment (dgrammatiko) [reply administrator/language/en-GB/joomla.ini:222]: My idea was to deprecate them in the 5.x repo so they could be removed in 6.0. Anyways I've added the deprecations

File: layouts/joomla/form/field/color/simple.php (hunk 3 above)
Comment (dgrammatiko) [reply layouts/joomla/form/field/color/simple.php]: reverted

**Comment by dgrammatiko** [comment]
@laoneo any interest here or should I close this?

**Comment by laoneo** [comment]
I'm surely interested, but not involved anymore in 6.0.

**Comment by dgrammatiko** [comment]
Ok, I might reopen this when there's a RM for 6

**Comment by Bodge-IT** [comment]
There's interest here! Were you thinking about any other fields?

**Comment by dgrammatiko** [comment]
Reopened here: https://github.com/joomla/joomla-cms/pull/45004

**Pull request closed without merging**